osquery/logger/plugins/kafka_producer.cpp

```diff
@@ -0,0 +1,236 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <unistd.h>
+
+#include <future>
+#include <memory>
+
+#include <librdkafka/rdkafka.h>
+
+#include <osquery/flags.h>
+#include <osquery/logger.h>
+
+namespace osquery {
+
+FLAG(string,
+     logger_kafka_brokers,
+     "localhost",
+     "Bootstrap broker(s) as a comma-separated list of host or host:port "
+     "(default port 9092)");
+
+FLAG(string,
+     logger_kafka_topic,
+     "osquery",
+     "Kafka topic to publish logs under");
+
+FLAG(string,
+     logger_kafka_acks,
+     "all",
+     "The number of acknowledgments the Kafka producer requires the leader to "
+     "have received before considering a request complete.  Valid values are "
+     "'0', '1', and 'all'");
+
+void delKafkaHandle(rd_kafka_t* k) {
+  rd_kafka_destroy(k);
+};
+
+void delKafkaTopic(rd_kafka_topic_t* kt) {
+  rd_kafka_topic_destroy(kt);
+};
+
+class KafkaProducerPlugin : public LoggerPlugin {
+ public:
+  /*
+   * @brief Logs string s as payload to configured Kafka brokers.
+   *
+   * Calls rd_kafka_poll regardless of production success.  rd_kafka_poll
+   * invokes callback that actually reports if message sends failed.
+   */
+  Status logString(const std::string& s) override;
+
+  /**
+   * @brief Initializes the Kafka producer.
+   *
+   * Setups producer with necessary configurations for interacting with Kafka
+   * brokers.  Registers the os hostname as the Kafka client.id.
+   */
+  void init(const std::string& name,
+            const std::vector<StatusLogLine>& log) override;
+
+  /**
+   * @brief Flushes final messages.
+   *
+   * Checks if Kafka producer is running and if so, flushes remaining messages
+   * to the brokers waiting a max of 3 seconds.
+   */
+  void stop();
+
+  KafkaProducerPlugin() : running_(false) {}
+  ~KafkaProducerPlugin();
+
+  KafkaProducerPlugin(KafkaProducerPlugin const&) = delete;
+  KafkaProducerPlugin& operator=(KafkaProducerPlugin const&) = delete;
+
+ private:
+  /// Smart pointer to the Kafka producer.
+  std::unique_ptr<rd_kafka_t, std::function<void(rd_kafka_t*)>> producer_;
```

Comment:
You'll want a `Mutex` (see `osquery/core.h`) and some well-scoped `WriteLock`'s to protect this as it is used in your async and in the class's dtor/etc.